File: add-binary/Solution67.java
// https://leetcode.com/problems/add-binary/

class Solution67 {

    // T: O(n), S: O(n)
    public String addBinary(String a, String b) {
        int i = a.length() - 1, j = b.length() - 1;
        int carry = 0;
        var sb = new StringBuilder();
        while (i >= 0 || j >= 0) {
            int x = (i >= 0) ? a.charAt(i) - '0' : 0;
            int y = (j >= 0) ? b.charAt(j) - '0' : 0;
            int sum = x + y + carry;
            int digit = sum % 2;
            sb.append(digit);
            carry = sum / 2;
            i--;
            j--;
        }
        if (carry > 0) {
            sb.append(carry);
        }
        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        var sol = new Solution67();
        var s1 = "11";
        var s2 = "1";
        System.out.println(sol.addBinary(s1, s2)); // 100
    }

}

File: add-binary/solution67.py
# https://leetcode.com/problems/add-binary/

class Solution67:

    # T: O(n), S: O(n)
    def addBinary(self, a: str, b: str) -> str:
        i, j = len(a) - 1, len(b) - 1
        carry = 0
        result = []
        while i >= 0 or j >= 0:
            x = int(a[i]) if i >= 0 else 0
            y = int(b[j]) if j >= 0 else 0
            sum = x + y + carry
            digit = sum % 2
            result.append(str(digit))
            carry = sum // 2
            i -= 1
            j -= 1
        if carry > 0:
            result.append(str(carry))
        return ''.join(result[::-1])


def main() -> None:
    sol = Solution67()
    s1 = "11"
    s2 = "1"
    print(sol.addBinary(s1, s2))  # 100


if __name__ == "__main__":
    main()
